pkg/kubelet/cm/devicemanager/topology_hints.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package devicemanager
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/socketmask"
+)
+
+// GetTopologyHints implements the TopologyManager HintProvider Interface which
+// ensures the Device Manager is consulted when Topology Aware Hints for each
+// container are created.
+func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) map[string][]topologymanager.TopologyHint {
+	deviceHints := make(map[string][]topologymanager.TopologyHint)
+
+	for resourceObj, requestedObj := range container.Resources.Limits {
+		resource := string(resourceObj)
+		requested := int(requestedObj.Value())
+
+		if m.isDevicePluginResource(resource) {
+			if aligned := m.deviceHasTopologyAlignment(resource); !aligned {
+				klog.Infof("[devicemanager] Resource '%v' does not have a topology preference", resource)
+				deviceHints[resource] = nil
+				continue
+			}
+
+			available := m.getAvailableDevices(resource)
+			if available.Len() < requested {
+				klog.Errorf("[devicemanager] Unable to generate topology hints: requested number of devices unavailable for '%s': requested: %d, available: %d", resource, requested, available.Len())
+				deviceHints[resource] = []topologymanager.TopologyHint{}
+				continue
+			}
+
+			deviceHints[resource] = m.generateDeviceTopologyHints(resource, available, requested)
+		}
+	}
+
+	return deviceHints
+}
+
+func (m *ManagerImpl) deviceHasTopologyAlignment(resource string) bool {
```

Comment:
A discussion is going on here https://github.com/intel/sriov-network-device-plugin/pull/160#discussion_r316146230 about whether these are the right semantics or not.